proposal/package-importer.d.ts.md

```diff
@@ -0,0 +1,386 @@
+packages, this will be the primary method used for the Node package importer. We
+will support both the `"sass"` and the `"style"` conditions, as Sass can also
+use the CSS exports exposed through `"style"`. While in practice, `"style"`
+tends to be used solely for `css` files, we will support `scss`, `sass` and
+`css` files for either `"sass"` or `"style"`.
+
+## Types
+
+```ts
+import '../spec/js-api';
+```
+
+#### `usePkgImporter`
+
+If true, the compiler will use the built-in package importer to resolve any url
+with the `pkg` scheme. This importer follows Node logic to locate Sass files.
+
+Defaults to false.
+
+```ts
+declare module '../spec/js-api/options' {
+  interface Options<sync extends 'sync' | 'async'> {
+    /**
+     * Whether or not to enable the built-in package importer to resolve any url
+     * with the `pkg` scheme. This importer follows Node.js resolution logic.
+     *
+     * @defaultValue `false`
+     * @category Input
+     */
+    usePkgImporter?: boolean;
+  }
+}
+```
+
+## Semantics
+
+### Package Importers
+
+This proposal defines the requirements for Package Importers written by users or
+provided by implementations. It is a type of [Filesystem Importer] and will
+handle non-canonical URLs:
+
+- with the scheme `pkg`
+- followed by a package name
+- optionally followed by a path, with path segments separated with a forward
+  slash.
+
+The package name will often be the first path segment, but the importer should
+take into account any conventions in the environment. For instance, Node
+supports scoped package names, which start with `@` followed by 2 path segments.
+
+[filesystem importer]: ../spec/modules.md#filesystem-importer
+
+### Node Specific Semantics
+
+The Node package importer is an [importer] with an associated absolute `pkg:`
+URL named `base`. When the Node package importer is invoked with a string named
+`string` and a [previous url] `previousUrl`:
+
+- Let `url` be the result of [parsing `string` as a URL][parsing a url] with
+  `base` as the base URL. If this returns a failure, throw that failure.
+- If `url`'s scheme is not `pkg`, return null.
+- Let `resolved` be the result of [resolving a `pkg` URL] with `url` and `previousURL`
+- If `resolved` is null, return null.
+- Let `text` be the contents of the file at `resolved`.
+- Let `syntax` be:
+  - "scss" if `url` ends in `.scss`.
+  - "indented" if `url` ends in `.sass`.
+  - "css" if `url` ends in `.css`.
+  > The algorithm for [resolving a `file:` URL](../spec/modules.md#resolving-a-file-url)
+  > guarantees that `url` will have one of these extensions.
+- Return `text`, `syntax`, and `resolved`.
+
+[parsing a URL]: https://url.spec.whatwg.org/#concept-url-parser
+[resolving a `pkg` URL]: #resolving-a-pkg-url
+
+> Note that this algorithm does not automatically resolve index files, partials
+> or extensions, except where specified. When using the `pkg:` url scheme,
+> authors need to explicitly use the fully resolved filename.
```

Comment:
I understand requiring package authors to explicitly list their entrypoints, but I think the load URLs should be able to write `@use "pkg:bootstrap/variables"` and not `@use "pkg:bootstrap/_variables.scss"`.

Reply:
@nex3 To tease out the issue a bit, this is a place where the 2 file resolution systems don't quite match up. 

The question is for URLs with subpaths where the subpath refers to an entrypoint alias, and not to a filepath on the filesystem. 

Take a `package.json` with the following in exports-

```json
{
  "exports": {
    "./_variables.scss": {
      "sass": "./src/sass/_libraryVariables.scss"
    }
  }
}
```

Should we allow the package consumer to write `@use 'pkg:packageName/variables'` in this case? Or would we expect the library author to set the entrypoint alias to `variables`? The author could also choose to expose both. The [Node docs](https://nodejs.org/api/packages.html#package-entry-points) recommend being exhaustive in explicitly naming all subpaths that should resolve to a specific file.

The Node Exports section essentially creates a parallel file alias structure which does not necessarily align with the real file system. The current procedure fuzzy matches only when accessing files in the real file system, but not via the Node exports aliases. 

The other implication of this is that if the package author has a export for `_index.scss` defined as an alias to a nested file, the step that checks for an `index` file at the package root will not find that when a package author uses `@use "pkg:packageName"`.  It would work for them to add an export for `.`, however.

Should we remove this differentiation, and do fuzzy matching for the aliases that the library users have set up as well? 

In the meantime, I've updated this comment to better explain the behavior-

```
> Note that this algorithm does not automatically resolve index files, partials
> or extensions when resolving paths defined in the package.json `exports`
> section. When using conditional exports, package authors need to explicitly
> expose all paths that should resolve to a Sass file.
```